src/map/ai/controllers/trust_controller.cpp

```diff
@@ -118,23 +142,61 @@ void CTrustController::DoCombatTick(time_point tick)
 
 void CTrustController::DoRoamTick(time_point tick)
 {
-    if (POwner->PMaster->PAI->IsEngaged() && GetTopEnmity() == POwner->PMaster)
+    auto PMaster = static_cast<CCharEntity*>(POwner->PMaster);
+    bool trustEngageCondition = (PMaster->GetBattleTarget() && PMaster->GetBattleTarget()->PLastAttacker == PMaster);
+
+    if (PMaster->PAI->IsEngaged() && trustEngageCondition)
     {
-        POwner->PAI->Internal_Engage(POwner->PMaster->GetBattleTargetID());
+        POwner->PAI->Internal_Engage(PMaster->GetBattleTargetID());
     }
 
-    float currentDistance = distance(POwner->loc.p, POwner->PMaster->loc.p);
+    uint8 currentPartyPos = GetPartyPosition();
+    CBattleEntity* PFollowTarget = (GetPartyPosition() > 0) ? (CBattleEntity*)PMaster->PTrusts.at(currentPartyPos - 1) : POwner->PMaster;
+    float currentDistance = distance(POwner->loc.p, PFollowTarget->loc.p);
+
+    for (auto POtherTrust : PMaster->PTrusts)
+    {
+        if (POtherTrust != POwner && distance(POtherTrust->loc.p, POwner->loc.p) < 1.0f)
+        {
+            auto angle = getangle(POwner->loc.p, POtherTrust->loc.p) + 64;
+            auto amount = (currentPartyPos % 2) ? 1.0f : -1.0f;
+            position_t new_pos{ POwner->loc.p.x - (cosf(rotationToRadian(angle)) * amount),
+                POtherTrust->loc.p.y, POwner->loc.p.z + (sinf(rotationToRadian(angle)) * amount), 0, 0 };
+
+            if (POwner->PAI->PathFind->ValidPosition(new_pos))
+            {
+                POwner->PAI->PathFind->PathAround(new_pos, RoamDistance, PATHFLAG_RUN | PATHFLAG_WALLHACK);
+                POwner->PAI->PathFind->FollowPath();
+            }
+            break;
+        }
+    }
 
     if (currentDistance > RoamDistance)
     {
-        auto new_position = GetDeclumpedPosition(POwner->PMaster->loc.p);
-        if (currentDistance < 30.0f && POwner->PAI->PathFind->PathAround(new_position, 2.0f, PATHFLAG_RUN | PATHFLAG_WALLHACK | PATHFLAG_SLIDE))
+        if (currentDistance < WarpDistance && POwner->PAI->PathFind->PathAround(PFollowTarget->loc.p, RoamDistance, PATHFLAG_RUN | PATHFLAG_WALLHACK))
         {
             POwner->PAI->PathFind->FollowPath();
         }
         else if (POwner->GetSpeed() > 0)
         {
-            POwner->PAI->PathFind->WarpTo(POwner->PMaster->loc.p, RoamDistance);
+            POwner->PAI->PathFind->WarpTo(PFollowTarget->loc.p, RoamDistance);
+        }
+    }
+
+    if (POwner->CanRest() &&
+        m_Tick - m_CombatEndTime > 10s &&
```

Comment:
A thought occurred that we might be able to cheat by setting a special regen status effect on disengage - which is cleared on engage - and then letting the status effect container handle it.

I don't know if it'd be _better_ than doing it yourself here; just bringing it up as an option in case someone else might say, "Yes, that'd be better because X and Y".

Reply:
I think it would probably be better to do the sneaky regen, this system I'm using has the healing "ticks" not in line with regular healing ticks

Reply:
Hmmm, I tried it out with a regen/refresh status and I couldn't get it to work as cleanly as I'd like, I'll come back to this later if needed